### ISSUE TITLE
False positive `invalid-method-override` when overriding overloaded method with matching @overload declarations

### ISSUE THREAD

**Issue by gaborbernat** [astral-sh/ty#3389]
### Summary

ty emits a false-positive `invalid-method-override` when a subclass overrides an overloaded method and declares matching `@overload` declarations. ty checks the **implementation** signature against the parent's overloads instead of checking the declared `@overload` signatures. mypy and pyright are both silent on the same code.

### Minimal repro

```python
from argparse import ArgumentParser, Namespace
from typing import TypeVar, overload
from collections.abc import Sequence

_N = TypeVar("_N", bound=Namespace)


class MyParser(ArgumentParser):
    @overload
    def parse_args(self, args: Sequence[str] | None = None, namespace: None = None) -> Namespace: ...
    @overload
    def parse_args(self, args: Sequence[str] | None, namespace: _N) -> _N: ...
    @overload
    def parse_args(self, *, namespace: _N) -> _N: ...
    def parse_args(self, args: Sequence[str] | None = None, namespace: Namespace | None = None) -> Namespace:
        return super().parse_args(args, namespace)
```

### Command

```
ty check repro.py
```

### ty output

```
error[invalid-method-override]: Invalid override of method `parse_args`
   --> repro.py:15:9
    |
 15 |     def parse_args(self, args: Sequence[str] | None = None, namespace: Namespace | None = None) -> Namespace:
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Definition is incompatible with `ArgumentParser.parse_args`
    |
   ::: stdlib/argparse.pyi:312:9
    |
312 |     def parse_args(self, *, namespace: _N) -> _N: ...
    |         ---------------------------------------- `ArgumentParser.parse_args` defined here
    |
info: This violates the Liskov Substitution Principle

Found 1 diagnostic
```

### Expected behaviour

No error. The subclass declares `@overload` signatures that exactly mirror the parent's, including the `(*, namespace: _N) -> _N` overload. The implementation body is a technical detail that cannot simultaneously satisfy all overloads with a single concrete return type — that is the entire reason `@overload` exists.

ty should check the declared `@overload` signatures against the parent's overloads, not the implementation signature.

### Cross-checks

- `mypy repro.py`: no issues
- `pyright repro.py`: no issues

### Real-world impact

This is blocking [tox-dev/tox#3930](https://github.com/tox-dev/tox/pull/3930) CI from passing. The tox codebase has valid overrides of `ArgumentParser.parse_args` and `ArgumentParser.parse_known_args` with correct `@overload` declarations, but ty flags the implementations as invalid overrides.

### Version

```
ty 0.0.33
```

**Comment by charliermarsh** [comment]
It seems like [Mypy](https://mypy-play.net/?gist=b9e68aba0b7d1c925e9a6264fa6023ad) and [Pyright](https://pyright-play.net/?strict=true&code=GYJw9gtgBAhiDmAHOBnAplAlhRYQBcoBBBAVwjQDt8AFVNEAGigDkYKVkBjNAKFEhR8AT0SZK8LDjyEAKqLQA1OMzAA3BgBswMACb9w0LmE2a0XfJjCUUAOhgAjLlNwEoAZTQBHUlR69eAH0WKABeKHlEJTgACgAiYLjmBzBSSl1Qtg5uNABKAN4uTRgUFCgAWWE6EHQQGJJ4cipaehBcgC5eKG6oAAF1LR19HqhdNGAoZBq0QLh4FBj0TWBmOZR2j29fSh4AbRR8EABdKAAfVmsMcJZL5kp2NE4YHg2byiuL99yoAFoAPlYDyeLygtjBXR6-Q0IG0egh3TGEym6FmCAWSxWsDRG08Pj8aH2hxO5zeaDuQJyG2C33%2BUGCGzBtnhfQGMKGzMRk3oqPmizQy2YACpydlnmgqSwaQD6aDwSNOciZms%2BQKsfMcVt8YTjmdPh9SSLHpTAaKeLrSWE9VKTUaxZ0RiMQGh8KQQJQoChSFE6rlbIqeQs1obgXleEA) both error on this?

**Comment by gaborbernat** [comment]
Let me check 🤔

**Comment by carljm** [comment]
We don't check the implementation against the base overloads, we are definitely checking overloaded signature against overloaded signature. Our diagnostics here could be a lot better, though; they point at the implementation which kind of suggests the wrong thing.

I think you are running into https://github.com/python/typeshed/pull/15613, which got pulled into ty 0.0.33. Your overloads now need to accept `Iterable[str]`, not `Sequence[str]`. We are fine with this:

```py
from argparse import ArgumentParser, Namespace
from typing import TypeVar, overload, Any, Iterable

_N = TypeVar("_N", bound=Namespace)


class MyParser(ArgumentParser):
    @overload
    def parse_args(self, args: Iterable[str] | None = None, namespace: None = None) -> Namespace: ...
    @overload
    def parse_args(self, args: Iterable[str] | None, namespace: _N) -> _N: ...
    @overload
    def parse_args(self, *, namespace: _N) -> _N: ...
    def parse_args(self, args: Iterable[str] | None = None, namespace: Namespace | None = None) -> Namespace:
        return super().parse_args(args, namespace)
```

I think there's another issue here that mypy/pyright highlight, which is that the stdlib overloads use an unbounded typevar, whereas your override uses a typevar bound to `Namespace`, which means your override can't handle all the same inputs the base method can. It looks like we currently don't error on that, but we should.

**Comment by carljm** [comment]
#2154 is our open issue tracking better Liskov diagnostics for overriding overloaded methods.

**Comment by carljm** [comment]
And https://github.com/astral-sh/ty/issues/3114 is the issue tracking the fact that we ought to error based on the differing upper bound.

I think beyond those two issues, the only problem here is https://github.com/python/typeshed/pull/15613, so closing this as non-actionable.

Thanks for the report!
